File: app/src/main/java/com/parabola/newtone/mvp/presenter/TabArtistPresenter.java
package com.parabola.newtone.mvp.presenter;

import com.parabola.domain.executor.SchedulerProvider;
import com.parabola.domain.interactor.ArtistInteractor;
import com.parabola.domain.model.Track;
import com.parabola.domain.repository.ArtistRepository;
import com.parabola.domain.repository.SortingRepository;
import com.parabola.domain.repository.TrackRepository;
import com.parabola.domain.settings.ViewSettingsInteractor;
import com.parabola.newtone.di.app.AppComponent;
import com.parabola.newtone.mvp.view.TabArtistView;
import com.parabola.newtone.ui.router.MainRouter;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.internal.functions.Functions;
import io.reactivex.internal.observers.ConsumerSingleObserver;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public final class TabArtistPresenter extends MvpPresenter<TabArtistView> {

    private static final String TAG = TabArtistPresenter.class.getSimpleName();

    @Inject MainRouter router;

    @Inject ArtistInteractor artistInteractor;
    @Inject SortingRepository sortingRepo;
    @Inject TrackRepository trackRepo;
    @Inject ViewSettingsInteractor viewSettingsInteractor;

    @Inject SchedulerProvider schedulers;

    private final CompositeDisposable disposables = new CompositeDisposable();

    public TabArtistPresenter(AppComponent appComponent) {
        appComponent.inject(this);
    }

    @Override
    protected void onFirstViewAttach() {
        disposables.addAll(
                observeAllArtists(),
                observeAllArtistsSorting(),
                observeTrackItemViewUpdates(),
                observeIsItemDividerShowed(),
                observeTrackDeleting());
    }

    @Override
    public void onDestroy() {
        disposables.dispose();
    }


    private Disposable observeAllArtists() {
        return artistInteractor.observeAllArtistsUpdates()
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui())
                .subscribe(getViewState()::refreshArtists);
    }


    private Disposable observeAllArtistsSorting() {
        return sortingRepo.observeAllArtistsSorting()
                //включаем/отключаем показ секции в списке, если отсортирован по имени
                .map(sorting -> sorting == ArtistRepository.Sorting.BY_NAME)
                .subscribe(getViewState()::setSectionShowing);
    }

    private Disposable observeTrackItemViewUpdates() {
        return viewSettingsInteractor.observeArtistItemViewUpdates()
                .subscribe(getViewState()::setItemViewSettings);
    }

    private Disposable observeIsItemDividerShowed() {
        return viewSettingsInteractor.observeIsItemDividerShowed()
                .subscribe(getViewState()::setItemDividerShowing);
    }

    private Disposable observeTrackDeleting() {
        return trackRepo.observeTrackDeleting()
                .flatMapSingle(deletedTrackId -> artistInteractor.getAll())
                .observeOn(schedulers.ui())
                .subscribe(getViewState()::refreshArtists);
    }

    public void onItemClick(int artistId) {
        router.openArtist(artistId);
    }


    public void onClickMenuShuffle(int artistId) {
        artistInteractor.shuffleArtist(artistId);
    }

    public void onClickMenuAddToPlaylist(int artistId) {
        trackRepo.getByArtist(artistId)
                .flatMapObservable(Observable::fromIterable)
                .map(Track::getId)
                .toList()
                .map(ids -> ids.stream().mapToInt(Integer::intValue).toArray())
                .subscribe(new ConsumerSingleObserver<>(
                        router::openAddToPlaylistDialog,
                        Functions.ERROR_CONSUMER
                ));
    }
}

File: app/src/main/java/com/parabola/newtone/presentation/main/artists/TabArtistPresenter.kt
package com.parabola.newtone.presentation.main.artists

import com.parabola.domain.interactor.ArtistInteractor
import com.parabola.domain.model.Track
import com.parabola.domain.repository.ArtistRepository
import com.parabola.domain.repository.SortingRepository
import com.parabola.domain.repository.TrackRepository
import com.parabola.domain.settings.ViewSettingsInteractor
import com.parabola.newtone.di.app.AppComponent
import com.parabola.newtone.presentation.base.BasePresenter
import com.parabola.newtone.presentation.router.MainRouter
import io.reactivex.Observable
import moxy.InjectViewState
import javax.inject.Inject

@InjectViewState
class TabArtistPresenter(
    appComponent: AppComponent,
) : BasePresenter<TabArtistView>() {

    @Inject
    lateinit var router: MainRouter

    @Inject
    lateinit var artistInteractor: ArtistInteractor

    @Inject
    lateinit var sortingRepo: SortingRepository

    @Inject
    lateinit var trackRepo: TrackRepository

    @Inject
    lateinit var useCases: TabArtistScreenUseCases

    @Inject
    lateinit var viewSettingsInteractor: ViewSettingsInteractor

    init {
        appComponent.inject(this)
    }


    override fun onFirstViewAttach() {
        observeAllArtists()
        observeAllArtistsSorting()
        observeTrackItemViewUpdates()
        observeIsItemDividerShowed()
        observeTrackDeleting()
    }

    private fun observeAllArtists() {
        artistInteractor.observeAllArtistsUpdates()
            .schedule(onNext = viewState::refreshArtists)
    }

    private fun observeAllArtistsSorting() {
        sortingRepo.observeAllArtistsSorting() //включаем/отключаем показ секции в списке, если отсортирован по имени
            .map { it == ArtistRepository.Sorting.BY_NAME }
            .schedule(onNext = viewState::setSectionShowing)
    }

    private fun observeTrackItemViewUpdates() {
        viewSettingsInteractor.observeArtistItemViewUpdates()
            .schedule(onNext = viewState::setItemViewSettings)
    }

    private fun observeIsItemDividerShowed() {
        viewSettingsInteractor.observeIsItemDividerShowed()
            .schedule(onNext = viewState::setItemDividerShowing)
    }

    private fun observeTrackDeleting() {
        trackRepo.observeTrackDeleting()
            .flatMapSingle { artistInteractor.all }
            .schedule(onNext = viewState::refreshArtists)
    }


    fun onItemClick(artistId: Int) {
        router.openArtist(artistId)
    }

    fun onClickMenuShuffle(artistId: Int) {
        useCases.shuffleArtist(artistId)
            .schedule()
    }

    fun onClickMenuAddToPlaylist(artistId: Int) {
        trackRepo.getByArtist(artistId)
            .flatMapObservable { artistTracks -> Observable.fromIterable(artistTracks) }
            .map(Track::getId)
            .toList()
            .map { trackIds -> trackIds.toIntArray() }
            .schedule(onSuccess = router::openAddToPlaylistDialog)
    }

}
